image/preseed/preseed_linux.go

```diff
@@ -443,50 +503,63 @@ func prepareClassicChroot(preseedChroot string) (*targetSnapdInfo, func(), error
 		rootDir = "/"
 	}
 
-	coreSnapPath, _, err := systemSnapFromSeed(dirs.SnapSeedDirUnder(rootDir), "")
-	if err != nil {
-		return nil, nil, err
-	}
-
-	// create mountpoint for core/snapd
-	where := filepath.Join(rootDir, snapdMountPath)
-	if err := os.MkdirAll(where, 0755); err != nil {
-		return nil, nil, err
+	unmountProc := func() {}
+	if !osutil.FileExists("/proc/self/cmdline") {
+		cmd := exec.Command("mount", "-t", "proc", "none", "/proc")
+		if out, err := cmd.CombinedOutput(); err != nil {
+			return nil, nil, fmt.Errorf("Cannot mount proc and /proc is not available: %s\nOutput: %s", err, out)
+		}
+		unmountProc = func() {
+			cmd := exec.Command("umount", "/proc")
+			if out, err := cmd.CombinedOutput(); err != nil {
+				fmt.Fprintf(Stdout, "cannot unmount /proc: %s\noutput: %s\n", err, out)
+			}
+		}
 	}
 
-	removeMountpoint := func() {
-		if err := os.Remove(where); err != nil {
-			fmt.Fprintf(Stderr, "%v", err)
+	unmountDev := func() {}
+	if !osutil.FileExists("/dev/loop-control") {
+		cmd := exec.Command("mount", "-t", "devtmpfs", "none", "/dev")
+		if out, err := cmd.CombinedOutput(); err != nil {
+			unmountProc()
```

Comment:
an approach that we use when there is a growing set of undo/cleanup functions to call on error is to have a slice of functions and append to it and then have a common defer that  like:

```
defer func() {
   if err != nil {
       for  i := len(cleanups)-1; i >=0; i-- {
          cleanups[i]()
       }
   }
}()

```

or similar